control-plane/agents/core/src/core/scheduling/volume.rs

```diff
@@ -0,0 +1,235 @@
+        Self {
+            spec: create.into(),
+        }
+    }
+}
+impl From<&VolumeSpec> for GetSuitablePools {
+    fn from(spec: &VolumeSpec) -> Self {
+        Self { spec: spec.clone() }
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct GetSuitablePoolsContext {
+    pub(crate) registry: Registry,
+    spec: VolumeSpec,
+}
+
+impl Deref for GetSuitablePoolsContext {
+    type Target = VolumeSpec;
+
+    fn deref(&self) -> &Self::Target {
+        &self.spec
+    }
+}
+impl Deref for GetSuitablePools {
+    type Target = VolumeSpec;
+
+    fn deref(&self) -> &Self::Target {
+        &self.spec
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct IncreaseVolumeReplica {
+    context: GetSuitablePoolsContext,
+    list: Vec<PoolItem>,
+}
+
+impl IncreaseVolumeReplica {
+    pub(crate) async fn builder(request: impl Into<GetSuitablePools>, registry: &Registry) -> Self {
+        Self {
+            context: GetSuitablePoolsContext {
+                registry: registry.clone(),
+                spec: request.into().spec,
+            },
+            list: PoolItemLister::list(registry).await,
+        }
+    }
+    pub(crate) async fn builder_with_defaults(
+        request: impl Into<GetSuitablePools>,
+        registry: &Registry,
+    ) -> Self {
+        Self::builder(request, registry)
+            .await
+            // filter pools according to the following criteria (any order):
+            // 1. if allowed_nodes were specified then only pools from those nodes
+            // can be used.
+            // 2. pools should have enough free space for the
+            // volume (do we need to take into account metadata?)
+            // 3. ideally use only healthy(online) pools with degraded pools as a
+            // fallback
+            // 4. only one replica per node
+            .filter(NodeFilters::online_nodes)
+            .filter(NodeFilters::allowed_nodes)
+            .filter(NodeFilters::used_nodes)
+            .filter(PoolFilters::healthy_pools)
+            .filter(PoolFilters::enough_free_space)
+            // sort pools in order of preference (from least to most number of replicas)
+            .sort(PoolSorters::sort_by_replica_count)
+    }
+}
+
+#[async_trait::async_trait(?Send)]
+impl ResourceFilter for IncreaseVolumeReplica {
+    type Request = GetSuitablePoolsContext;
+    type Item = PoolItem;
+
+    fn filter<P: FnMut(&Self::Request, &Self::Item) -> bool>(mut self, mut filter: P) -> Self {
+        let request = self.context.clone();
+        self.list = self
+            .list
+            .into_iter()
+            .filter(|v| filter(&request, v))
+            .collect();
+        self
+    }
+    async fn filter_async<Fn, Fut>(mut self, mut filter: Fn) -> Self
+    where
+        Fn: FnMut(&Self::Request, &Self::Item) -> Fut,
+        Fut: Future<Output = bool>,
+    {
+        let mut self_clone = self.clone();
+        self.list.clear();
+
+        for i in self.list {
```

Comment:
You are clearing the list and then iterating over it?

Reply:
I'm removing this as it's not used anymore. Long live sync code :)